### PR TITLE
[BIOMAGE-1137] - Order samples as defined in project

### PR DESCRIPTION
**Pull request by aerlaut**
# Background
#### Link to issue 

https://biomage.atlassian.net/browse/BIOMAGE-1137

#### Link to staging deployment URL 

In UI repo

#### Links to any Pull Requests related to this

UI - https://github.com/biomage-ltd/ui/pull/408
API - https://github.com/biomage-ltd/api/pull/188

#### Anything else the reviewers should know about the changes here

# Changes
### Code changes

# Definition of DONE
Your changes will be ready for merging after each of the steps below have been completed:

### Testing
- [ ] Unit tests written
- [X] Tested locally with Inframock (with latest production data downloaded with biomage experiment pull)
- [X] Deployed to staging

To set up easy local testing with inframock, follow the instructions here: https://github.com/biomage-ltd/inframock
To deploy to the staging environment, follow the instructions here: https://github.com/biomage-ltd/biomage-utils

### Documentation updates
Is all relevant documentation updated to reflect the proposed changes in this PR?

- [ ] Relevant Github READMEs updated
- [ ] Relevant wiki pages created/updated

### Approvers
- [ ] Approved by a member of the core engineering team
- [ ] (UX changes) Approved by vickymorrison (this is her username, tag her if you need approval)

### Just before merging:
- [ ] After the PR is approved, the `unstage` script in here: https://github.com/biomage-ltd/biomage-utils is executed. This script cleans up your deployment to staging

### Optional
- [ ] Photo of a cute animal attached to this PR

### CODE REVIEW

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -249,7 +249,7 @@ jobs:
 
       - name: Change name of deployment file for staging deployment
         if:
-          github.event_name == 'push' && matrix.environment == 'staging'
+          (github.event_name == 'pull_request' || github.event_name == 'push') && matrix.environment == 'staging'
```
Comment (cosa65) [inline .github/workflows/ci.yaml]: what is this change? Doesn't seem related to your ticket.

File: src/api/route-services/gem2s.js [hunk 2]
```diff
@@ -49,7 +49,12 @@ class Gem2sService {
   static async generateGem2sTaskParams(experimentId) {
     const experiment = await (new ExperimentService()).getExperimentData(experimentId);
     const { samples } = await (new SamplesService()).getSamplesByExperimentId(experimentId);
-    const { metadataKeys } = await (new ProjectService()).getProject(experiment.projectId);
+    const {
+      metadataKeys,
+      samples: sampleIdsInOrder,
+    } = await new ProjectService().getProject(experiment.projectId);
+
+    console.log(sampleIdsInOrder);
```
Comment (cosa65) [inline src/api/route-services/gem2s.js]: Remove this consolelog

File: src/api/route-services/gem2s.js [hunk 3]
```diff
@@ -49,7 +49,12 @@ class Gem2sService {
   static async generateGem2sTaskParams(experimentId) {
     const experiment = await (new ExperimentService()).getExperimentData(experimentId);
     const { samples } = await (new SamplesService()).getSamplesByExperimentId(experimentId);
-    const { metadataKeys } = await (new ProjectService()).getProject(experiment.projectId);
+    const {
```
Comment (cosa65) [inline src/api/route-services/gem2s.js:52]: it re runs gem23s if we re order the samples and then click launch analysis, I think we could avoid this relatively easily.

In the api we would need to change the way we create the hash to check if we should re run or not the pipeline, one option could be by ordering the samples before generating the hash. (just ordering for the hash so that it isn’t order-sensitive, but not ordering the actual samples array that is going to be stored)

Comment (aerlaut) [reply src/api/route-services/gem2s.js:52]: Yup, this would fix the issue. The first time the project is launched after this update, it would re run GEM2S. However, on subsequent runs, it wouldn't.

File: src/api/route-services/gem2s.js (hunk 2 above)
Comment (aerlaut) [reply src/api/route-services/gem2s.js]: Removed

File: src/api/route-services/gem2s.js (hunk 3 above)
Comment (aerlaut) [reply src/api/route-services/gem2s.js:52]: It seems that the problem is bigger than this. Suggests moving this to another ticket.

Creating a separate variable to hold sample-order-invariant params as the source for hashing can bypass rerunning. However, the source of sample order in the UI is `cellSets`, and not rerunning GEM2S does not rearrange the order in cellSets. This necessitates other approaches such as: 

*Getting the list of experiment from `projects` in data-processing*
We don't yet have a function to get a single project, given an experimentId. Aside from that, loading projects depend on the currently logged in userId. This will be an issue down the line if we decided to support sharing of experiments.

*Storing the list of samples under an experiment object*
`experimentSettings.info` is updated everytime an experiment is launched. This object is used to store experiment configuration. We can easily pass the samples list to this object to be used later on down the line.

In data processing, sampleIds (`sampleKeys`) are derived from cellSets object. This [object actually stores the order the samples are inserted](https://www.stefanjudis.com/today-i-learned/property-order-is-predictable-in-javascript-objects-since-es2015/). Hence, we need to rewrite `sampleKeys` to get the data from `experimentSettings.info`.

This is all well for experiments that are launched from Data Management. However, if we refresh the page, `experimentSettings.info.samples` is missing. Upon refresh `experimentSettings.info` is populated by a server-side (`utils/ssr/getExperimentInfo`). The call `getExperimentInfo` gets data from the `experiments` table. Right now, we don't have any record of `sampleIds` inside the `samples` table. Hence, the easiest way is to save `sampleIds` into `experiments` before running `GEM2S`.

This can be done, but to me it seems to be a "hacky" fix, which might cause problems down the line. Hence, I'd rather this problem be moved to another ticket and given more thought.

File: .github/workflows/ci.yaml (hunk 1 above)
Comment (aerlaut) [reply .github/workflows/ci.yaml]: This was a change I made so that this branch compiled. It's already separated into PR #189

**Review by cosa65 — APPROVED** [review]

File: src/api/route-services/gem2s.js [hunk 4]
```diff
@@ -76,7 +79,15 @@ class Gem2sService {
...
-    return taskParams;
+    const orderInvariantSampleIds = [...sampleIdsInOrder].sort();
+
+    const hashParams = {
+      sampleIds: orderInvariantSampleIds,
+      sampleNames: orderInvariantSampleIds.map((sampleId) => samples[sampleId].name),
+      metadata: taskParams.metadata,
+    };
+
+    return { taskParams, hashParams };
```
Comment (cosa65) [inline src/api/route-services/gem2s.js:92]: I think we could add some unit testing to  this part at least to make sure that in the future we don't break this dependency on what causes it to run and what doesn't

**Review by cosa65 — CHANGES_REQUESTED** [review]

**Review by cosa65 — APPROVED** [review]
Approved so you can merge tomorrow, just let me know if you feel like anotehr review would be necessary.

File: src/api/general-services/pipeline-manage/index.js [hunk 5]
```diff
@@ -157,11 +155,7 @@ const createQCPipeline = async (experimentId, processingConfigUpdates) => {
   const accountId = await config.awsAccountIdPromise;
   const roleArn = `arn:aws:iam::${accountId}:role/state-machine-role-${config.clusterEnv}`;
   logger.log(`Fetching processing settings for ${experimentId}`);
-  const { processingConfig } = await experimentService.getProcessingConfig(experimentId);
-
-  const { samples } = await samplesService.getSamplesByExperimentId(experimentId);
-
-  const sampleIds = Object.keys(samples);
+  const { processingConfig, sampleIds } = await experimentService.getExperimentAttributes(experimentId, ['processingConfig', 'sampleIds']);
```
Comment (aerlaut) [inline src/api/general-services/pipeline-manage/index.js]: Create a wrapper function - for now - to make it uniform like the others.

File: src/api/route-services/experiment.js [hunk 6]
```diff
@@ -70,7 +94,7 @@ class ExperimentService {
 
     const marshalledData = convertToDynamoDbRecord({
       ':experimentName': body.name,
-      ':createdAt': body.createdAt,
+      ':createdDate': body.createdDate,
```
Comment (aerlaut) [inline src/api/route-services/experiment.js:77]: We'll need to make a migration for this

Comment (aerlaut) [reply src/api/route-services/experiment.js:77]: An easier solution is to just copy `createdAt` to `createdDate` in the migration

File: tests/api/route-services/gem2s.test.js [hunk 7]
```diff
@@ -183,4 +187,138 @@ describe('gem2sShouldRun', () => {
...
+        .mockImplementationOnce(() => Promise.resolve({ metadataKeys: [] })),
+    }));
+
+
+    const { taskParams: task2, hashParams: hash2 } = await Gem2sService.generateGem2sParams('experimentId');
+
+    expect(hash1).toEqual(hash2);
+  });
+
+  it('generateGem2sTaskParams returns different hash param if passed diferent information on vital fields', async () => {
```
Comment (aerlaut) [inline tests/api/route-services/gem2s.test.js]: Splint into 3 different tests

File: src/api/route-services/gem2s.js (hunk 4 above)
Comment (aerlaut) [reply src/api/route-services/gem2s.js:92]: Written

File: src/api/general-services/pipeline-manage/index.js (hunk 5 above)
Comment (aerlaut) [reply src/api/general-services/pipeline-manage/index.js]: created

File: tests/api/route-services/gem2s.test.js (hunk 7 above)
Comment (aerlaut) [reply tests/api/route-services/gem2s.test.js]: split

File: src/api/route-services/experiment.js (hunk 6 above)
Comment (aerlaut) [reply src/api/route-services/experiment.js:77]: created, now testing
